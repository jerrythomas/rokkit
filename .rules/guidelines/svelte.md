# Guidelines

- Use Svelte 5 runes mode for code
- Use SvelteKit for app
- Use UnoCSS for styling
- Use vitest for unit testing
- Use Playwright for e2e testing
- E2E tests should follow the page object pattern

## Svelte 5
{{[Svelte & Svelte Kit Distilled](https://svelte-llm.khromov.se/svelte-complete-distilled)}}
{{[svelte 5 documentation](https://svelte.dev/docs/svelte/llms.txt)}}

## Snippets

### Runes

Runes are special functions that replace reactivity declarations from Svelte 4.

```svelte
<script>
  // State
  let count = $state(0);

  // Derived values
  let doubled = $derived(count * 2);

  // Props
  let {
    name = 'world',
    value = $bindable(null)
  } = $props();

  // Effects
  $effect(() => {
    console.log(`Count changed to ${count}`);
  });
</script>
```

### Defining and passing snippets to components:

```svelte
<script>
  let items = ['one','two', 'three']
</script>
<!-- a header snippet -->
{#snippet header()}
  <h1>Custom header</h1>
{/snippet}

<!-- a snippet with parameters -->
{#snippet item(item, selected)}
  <div class:selected={selected}>
    {item.name}
  </div>
{/snippet}

<List {items} {header} {item} />
```

### Using snippets in a component:

```svelte
<script>
  let { header, item } = $props();
</script>

<div class="header">
  {@render header?.()}
</div>

{#each items as currentItem}
  {@render item(currentItem, selectedItem === currentItem)}
{/each}
```

To pass a snippet to another component:

#### ChildComponent.svelte

```svelte
<script>
  let { header } = $props();
</script>

{#if header}
  {@render header()}
{/if}
```

#### Snippets can be passed as properties to child component

```svelte
<script>
   import  ChildComponent from './ChildComponent.svelte'
</script>
{#snippet header()}
  <h1>Custom Header</h1>
{/snippet}

<ChildComponent {header} />
```

### Snippets can be enclosed within the child component.

```svelte
<script>
   import  ChildComponent from './ChildComponent.svelte'
</script>
<ChildComponent>
  {#snippet header()}
    <h1>Custom Header</h1>
  {/snippet}
</ChildComponent>
```
Both of these options work in a similar way

### Important: What NOT to do

Do not confuse snippets with React-style JSX or string templates:

```svelte
<!-- ❌ INCORRECT - Using arrow functions returning strings -->
<List
  header={() => `<h1>My List</h1>`}
  item={({ item }) => `<div>${item.name}</div>`}
/>

<!-- ✅ CORRECT - Using properly defined snippets -->
<!-- snippets are defined in the html section not in scripts block -->
{#snippet header()}
  <h1>My List</h1>
{/snippet}

{#snippet itemTemplate(item, selected)}
  <div class:selected={selected}>
    {item.name}
  </div>
{/snippet}

<List {header} item={itemTemplate} />
```

## Component Design Guidelines

- should be modular and data driven
- should be accessibile
- Reduced motion should be considered

## Project Structure

If the source code uses runes or the test uses runes the file extension should be `.svelte.js` or `.spec.svelte.js` respectively.

```
project/
│
├── src/
│   ├── index.js
│   ├── Component.svelte
│   └── utils.svelte.js
│
├── spec/
│   ├── index.spec.js
│   ├── Component.spec.js
│   └── utils.spec.svelte.js
├── e2e/
│   ├── home.e2e.js
│   └── login.e2e.js
└── README.md
```
