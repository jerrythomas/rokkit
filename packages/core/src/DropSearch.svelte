<script>
	import { defaultFields } from './constants.js'
	import List from './List.svelte'
	import Slider from './Slider.svelte'
	import { Item } from './items'

	export let data
	export let value = null
	export let fields = defaultFields
	export let using = { default: Item }

	$: using = { default: Item, ...using }
	$: fields = { ...defaultFields, ...fields }

	let opened = false
	let search
	let searchBox

	// on search change filter list
	function handleSelect(event) {
		console.log('selected', event)
		value = event.detail
		search = event.detail[fields.text]
	}
</script>

<!-- svelte-ignore a11y-no-noninteractive-tabindex -->
<div
	class="flex flex-col outline-none w-full h-12 relative dropdown"
	tabindex={0}
>
	<div class="flex flex-shrink-0 h-12 items-center pl-4 selected-item">
		<span class="flex flex-grow">
			<input
				type="text"
				class="p-0 border-none bg-transparent w-full"
				bind:value={search}
				bind:this={searchBox}
				on:focus={() => {
					opened = true
					searchBox.select()
				}}
				on:blur={() => {
					opened = false
				}}
			/>
		</span>
		{#if opened}
			<icon class="selector-opened" />
		{:else}
			<icon class="selector-closed" />
		{/if}
	</div>

	{#if opened}
		<Slider>
			<List
				bind:items={data}
				bind:value
				{fields}
				{using}
				on:select={handleSelect}
				on:change
			/>
		</Slider>
	{/if}
</div>
