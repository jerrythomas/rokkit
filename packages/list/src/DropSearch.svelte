<script>
	import { Icon } from '@sparsh-ui/icons'
	import List from './List.svelte'
	import Slider from './Slider.svelte'
	// import DropDownIcon from './icons/DropDownIcon.svelte'

	export let data
	export let component = null
	export let selected = null
	export let key = 'id'
	export let selectedKey

	let opened = false
	let search
	let searchBox

	// on search change filter list
	function handleSelect(event) {
		console.log('selected', event)
		selected = event.detail
		search = event.detail.name
	}
</script>

<div
	class="flex flex-col outline-none w-full h-12 relative dropdown"
	tabindex={0}
>
	<div class="flex flex-shrink-0 h-12 items-center pl-4 selected-item">
		<span class="flex flex-grow">
			<input
				type="text"
				class="p-0 border-none bg-transparent w-full"
				bind:value={search}
				bind:this={searchBox}
				on:focus={() => {
					opened = true
					searchBox.select()
				}}
				on:blur={() => {
					opened = false
				}}
			/>
		</span>
		<Icon name="selector" />
	</div>

	{#if opened}
		<Slider>
			<List
				bind:items={data}
				{key}
				{component}
				bind:selected={selectedKey}
				on:select={handleSelect}
				on:change
			/>
		</Slider>
	{/if}
</div>
