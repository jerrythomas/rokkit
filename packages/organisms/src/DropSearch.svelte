<script>
	import { defaultFields } from '@rokkit/core'
	import List from './List.svelte'
	import { Slider } from '@rokkit/atoms'
	import { Item } from '@rokkit/molecules'

	export let items
	export let value = null
	/** @type {import('@rokkit/core').FieldMapping} */
	export let fields = defaultFields
	export let using = { default: Item }

	$: using = { default: Item, ...using }
	$: fields = { ...defaultFields, ...fields }

	let opened = false
	let search
	let searchBox

	// on search change filter list
	function handleSelect(event) {
		console.log('selected', event)
		value = event.detail
		search = event.detail[fields.text]
	}
</script>

<!-- svelte-ignore a11y-no-noninteractive-tabindex -->
<div
	class="dropdown relative h-12 w-full flex flex-col outline-none"
	tabindex={0}
>
	<div class="selected-item h-12 flex flex-shrink-0 items-center pl-4">
		<span class="flex flex-grow">
			<input
				type="text"
				class="w-full border-none bg-transparent p-0"
				bind:value={search}
				bind:this={searchBox}
				on:focus={() => {
					opened = true
					searchBox.select()
				}}
				on:blur={() => {
					opened = false
				}}
			/>
		</span>
		{#if opened}
			<icon class="selector-opened" />
		{:else}
			<icon class="selector-closed" />
		{/if}
	</div>

	{#if opened}
		<Slider>
			<List
				bind:items
				bind:value
				{fields}
				{using}
				on:select={handleSelect}
				on:change
			/>
		</Slider>
	{/if}
</div>
