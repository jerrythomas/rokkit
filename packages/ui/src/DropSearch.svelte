<script>
	import { defaultFields, getText } from '@rokkit/core'
	import Select from './Select.svelte'

	let { options, name, value, fields = defaultFields, ...restProps } = $props()

	let searchText = $state('')
	let searchBox
	let filtered = $state(options)

	fields = { ...defaultFields, ...fields }

	function applySearch(event) {
		searchText = event.target.value
		if (searchText) {
			filtered = options.filter((option) =>
				getText(option, fields).toLowerCase().includes(searchText.toLowerCase())
			)
		} else filtered = options
	}
	function handleSelect(event) {
		value = event.detail
		searchText = getText(event.detail, fields)
	}
</script>

<Select {name} bind:value bind:options={filtered} {...restProps} {fields} onselect={handleSelect}>
	<span class="flex flex-grow">
		<input
			type="text"
			class="w-full border-none bg-transparent p-0"
			bind:value={searchText}
			bind:this={searchBox}
			onchange={applySearch}
		/>
	</span>
</Select>
