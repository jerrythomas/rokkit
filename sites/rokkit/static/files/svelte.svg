<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M18.2301 11.1217C21.1087 6.47773 15.3383 1.6935 11.2372 4.37297L7.337 6.85957C5.44834 8.08961 4.62832 10.453 5.78004 12.8817C2.90038 17.2501 8.50365 22.3655 12.7729 19.6332L16.6732 17.1466C18.7952 15.7494 19.2821 13.3398 18.2301 11.1217V11.1217ZM8.46536 8.46761L12.3599 5.98242C14.9184 4.45077 17.8644 7.12896 16.8023 9.62435C16.1715 9.12318 15.4685 8.81503 14.9345 8.67627C15.1375 7.87006 14.2016 7.27202 13.5507 7.68841L9.64763 10.1651C8.68258 10.8321 9.71148 12.204 10.6653 11.6284L12.1527 10.6803C15.4905 8.79676 18.6495 13.3192 15.5448 15.5357L11.6502 18.0209C9.30336 19.5215 6.10922 17.1175 7.20785 14.379C7.79145 14.8207 8.41131 15.1456 9.07564 15.3271C8.92896 16.2531 9.86333 16.7033 10.4694 16.3234L14.3626 13.8382C15.3233 13.1178 14.292 11.814 13.3464 12.375L11.8533 13.3231C8.55362 15.2514 5.32106 10.6843 8.46536 8.46761Z" fill="black"/>
</svg>
