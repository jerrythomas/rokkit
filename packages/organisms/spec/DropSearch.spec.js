import { describe, expect, it, beforeEach } from 'vitest'
import { cleanup, render } from '@testing-library/svelte'

import DropSearch from '../src/DropSearch.svelte'

describe('DropSearch.svelte', () => {
	beforeEach(() => cleanup())

	it('should render', () => {
		const { container } = render(DropSearch)
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})
})
