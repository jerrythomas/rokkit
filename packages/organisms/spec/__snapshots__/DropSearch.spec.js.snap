// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DropSearch.svelte > should render 1`] = `
<div>
  <input-select
    class="flex flex-col relative open"
    role="listbox"
    tabindex="0"
  >
    <selected-item
      aria-selected="false"
      class="w-full flex items-center"
      role="option"
      tabindex="-1"
    >
      <span
        class="flex flex-grow"
      >
        <input
          class="w-full border-none bg-transparent p-0"
          type="text"
        />
      </span>
       
      <icon
        aria-label="opened"
        class="flex flex-shrink-0 items-center justify-center "
        role="img"
        tabindex="-1"
      >
        <i
          aria-hidden="true"
          class="selector-opened"
        />
      </icon>
    </selected-item>
     
    <scroll
      class="flex flex-col absolute w-full z-10 overflow-scroll menu"
      style="top: 0px; animation: __svelte_3328443287_0 50ms linear 0ms 1 both;"
    >
      <list
        aria-label="list"
        class="flex flex-col w-full flex-shrink-0 select-none list"
        role="listbox"
        tabindex="-1"
      >
         
        <item
          aria-selected="false"
          class="item"
          data-path="0"
          role="option"
        >
          
           
          <p>
            1
          </p>
          
           
        </item>
        <item
          aria-selected="false"
          class="item"
          data-path="1"
          role="option"
        >
          
           
          <p>
            2
          </p>
          
           
        </item>
        <item
          aria-selected="false"
          class="item"
          data-path="2"
          role="option"
        >
          
           
          <p>
            3
          </p>
          
           
        </item>
      </list>
    </scroll>
  </input-select>
</div>
`;

exports[`DropSearch.svelte > should work with custom components 1`] = `
<div>
  <input-select
    class="flex flex-col relative open"
    role="listbox"
    tabindex="0"
  >
    <selected-item
      aria-selected="false"
      class="w-full flex items-center"
      role="option"
      tabindex="-1"
    >
      <span
        class="flex flex-grow"
      >
        <input
          class="w-full border-none bg-transparent p-0"
          type="text"
        />
      </span>
       
      <icon
        aria-label="opened"
        class="flex flex-shrink-0 items-center justify-center "
        role="img"
        tabindex="-1"
      >
        <i
          aria-hidden="true"
          class="selector-opened"
        />
      </icon>
    </selected-item>
     
    <scroll
      class="flex flex-col absolute w-full z-10 overflow-scroll menu"
      style="top: 0px; animation: __svelte_3328443287_0 50ms linear 0ms 1 both;"
    >
      <list
        aria-label="list"
        class="flex flex-col w-full flex-shrink-0 select-none list"
        role="listbox"
        tabindex="-1"
      >
         
        <item
          aria-selected="false"
          class="item"
          data-path="0"
          role="option"
        >
          <span>
            Alpha
          </span>
           
        </item>
        <item
          aria-selected="false"
          class="item"
          data-path="1"
          role="option"
        >
          <span>
            Beta
          </span>
           
        </item>
        <item
          aria-selected="false"
          class="item"
          data-path="2"
          role="option"
        >
          <span>
            Charlie
          </span>
           
        </item>
      </list>
    </scroll>
  </input-select>
</div>
`;
